pkg/ipamd/datastore/data_store.go

```diff
@@ -883,8 +1187,56 @@ func (ds *DataStore) GetENIIPs(eniID string) ([]string, error) {
 	}
 
 	var ipPool = make([]string, 0, len(eni.IPv4Addresses))
-	for ip := range eni.IPv4Addresses {
+	for ip, addr := range eni.IPv4Addresses {
+		// /32 IPs from SIP or PD pool will be in eni.IPv4Addresses
+		// hence this has to be checked, IP belonging to a prefix can be
+		// a possibility because of PD enable/disable knob toggle
+		if !eni.IsPDenabled && addr.Prefix != "" {
+			ds.log.Debugf("IP %s belongs to prefix pool so do not account", ip)
+			continue
+		}
 		ipPool = append(ipPool, ip)
 	}
 	return ipPool, nil
 }
+
+// GetENIPrefixes returns the known (allocated & unallocated) ENI Prefixed.
+func (ds *DataStore) GetENIPrefixes(eniID string) ([]string, error) {
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	eni, ok := ds.eniPool[eniID]
+	if !ok {
+		return nil, errors.New(UnknownENIError)
+	}
+
+	var ipPool = make([]string, 0, len(eni.IPv4Prefixes))
+	for prefix, prefixData := range eni.IPv4Prefixes {
+		ip := (prefix + "/" + strconv.Itoa(prefixData.PrefixLen))
+		ipPool = append(ipPool, ip)
+	}
+	return ipPool, nil
+}
+
+func (ds *DataStore) GetFreePrefixes() int {
+	freePrefixes := 0
+	for _, other := range ds.eniPool {
+		for _, otherPrefixes := range other.IPv4Prefixes {
+			if otherPrefixes.UsedIPs == 0 {
+				freePrefixes++
+			}
+		}
+
+	}
+	return freePrefixes
+}
+
+func (ds *DataStore) CleanupDataStore(eniID string, ipv4 string, force bool, enableIpv4PrefixDelegation bool) error {
+	var err error
+	if !enableIpv4PrefixDelegation {
+		err = ds.DelIPv4AddressFromStore(eniID, ipv4, force)
```

Comment:
I'm a bit confused here - isn't what we're deleting a property of the address itself (ie: is it a secondary IP or PD IP?) and not a property of what mode we're running in?

iow: what if the address was attached to this instance as a secondaryIP, but vpc-cni was started with PD enabled.   
 Now we're trying to free that unused secondary IP.  What happens in this function?

Reply:
So if vpc-cni started with PD enabled and is not assigned, I free it up at nodeInit - https://github.com/jayanthvn/amazon-vpc-cni-k8s/blob/pd_dev/pkg/ipamd/ipamd.go#L428. If the pod is assigned associated with secondary IP then on delete, I go free the IP.

Reply:
CleanupDataStore with PD enabled will just cleanup prefixes and associated IPs in the prefix.

Reply:
> if vpc-cni started with PD enabled and is not assigned, I free it up at nodeInit

I think only unassigned IPs get freed at startup.  Pods that are still running while ipamd is restarted will continue running (good), and I _think_ when they eventually exit, we will reach this function with their secondary IP?

Reply:
Yes the running pod will continue but when they exit, I am force cleaning it up here - https://github.com/jayanthvn/amazon-vpc-cni-k8s/blob/pd_dev/pkg/ipamd/rpc_handler.go#L206.